README.md
![code-climate-logo_black-200](https://user-images.githubusercontent.com/18341459/47682820-32937480-db93-11e8-9d81-e5052a22453b.png)

# Code Climate Test Reporter

Code Climate's test reporter is a binary that works in coordination with codeclimate.com to report test coverage data. Once you've set up test coverage reporting you can:
* view test coverage reports for each file alongside quality metrics like complexity, duplication, and churn,
* toggle between viewing code issues and test coverage line-by-line in the same source listings,
* block PRs from being merged if they don't meet your team's standards for test coverage percentage.

Code Climate accepts test coverage data from virtually any location, including locally run tests or your continuous integration (CI) service, and supports a variety of programming languages and test coverage formats, including Ruby, JavaScript, Go, Python, PHP, Java, and more.

For installation instructions, check out our docs on [Configuring Test Coverage](https://docs.codeclimate.com/docs/configuring-test-coverage) and [Test Coverage Troubleshooting Tips](https://docs.codeclimate.com/docs/test-coverage-troubleshooting-tips).

To sign up for Code Climate, head [here](https://codeclimate.com/quality/pricing/).

# Versioning
The test reporter's current version is documented in [VERSIONING/VERSION](https://github.com/codeclimate/test-reporter/blob/master/VERSIONING/VERSION), following the [Semantic Versioning](https://semver.org/) convention.

# Binaries

## Download
The test reporter is distributed as a pre-built binary named cc-test-reporter. You can fetch the pre-built binary from the following URLs:

### Linux
- [codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64](https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64)
- [codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-linux-amd64](https://codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-linux-amd64)

### Linux ARM64
- [codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-arm64](https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-arm64)
- [codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-linux-arm64](https://codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-linux-arm64)

### Linux netcgo (recommended if you're using a VPN)
- [codeclimate.com/downloads/test-reporter/test-reporter-latest-netcgo-linux-amd64](https://codeclimate.com/downloads/test-reporter/test-reporter-latest-netcgo-linux-amd64)
- [codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-netcgo-linux-amd64](https://codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-netcgo-linux-amd64)

### OS X
- [codeclimate.com/downloads/test-reporter/test-reporter-latest-darwin-amd64](https://codeclimate.com/downloads/test-reporter/test-reporter-latest-darwin-amd64)
- [codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-darwin-amd64](https://codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-darwin-amd64)

### Windows

- [https://codeclimate.com/downloads/test-reporter/test-reporter-latest-windows-amd64](https://codeclimate.com/downloads/test-reporter/test-reporter-latest-windows-amd64)
- [https://codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-windows-amd64](https://codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-windows-amd64)

#### e.g
>```console
>$ curl -L -O https://codeclimate.com/downloads/test-reporter/test-reporter-0.10.1-darwin-amd64
>```

## Verifying binaries

Along with the binaries you can download a file with a SHA 256 checksum for the given version from the link shown below, or you can attach it to your clipboard from the [docs page](https://docs.codeclimate.com/docs/configuring-test-coverage#locations-of-pre-built-binaries).

- [codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-darwin-amd64.sha256](https://codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-darwin-amd64.sha256)

To download the file containing the checksum using `curl`:
#### e.g
>```console
>$ curl -L -O https://codeclimate.com/downloads/test-reporter/test-reporter-0.10.1-darwin-amd64.sha256
>```

To check that a downloaded file matches the checksum, run it through `shasum` with a command such as:

```console
$ grep test-reporter-${VERSION}-darwin-amd64  test-reporter-${VERSION}-darwin-amd64.sha256 | shasum -a 256 -c -
```

The GPG detached signature of SHA checksums can be download analogously from the following url:

- [codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-darwin-amd64.sha256.sig](https://codeclimate.com/downloads/test-reporter/test-reporter-${VERSION}-darwin-amd64.sha256.sig)

You can use it with `gpg` to verify the integrity of your downloaded checksum. You will first need to import
the GPG publick key. To import the key:

```console
$ gpg --keyserver  keys.openpgp.org --recv-keys 9BD9E2DD46DA965A537E5B0A5CBF320243B6FD85
```

Then use the following command to verify the file's signature.

```console
$ gpg --verify test-reporter-${VERSION}-darwin-amd64.sha256.sig test-reporter-${VERSION}-darwin-amd64.sha256
```

## Copyright

See the [LICENSE](https://github.com/codeclimate/test-reporter/blob/master/LICENSE).


CHANGELOG.md
# Change log

## v0.11.1 [(2023-3-24)](https://github.com/codeclimate/test-reporter/releases/tag/v0.11.1)

* [NEW] Release Windows binary and update envy dependency version [#511](https://github.com/codeclimate/test-reporter/pull/511)


## v0.11.0 [(2023-3-3)](https://github.com/codeclimate/test-reporter/releases/tag/v0.11.0)

* [NEW] Add support for .NET Dotcover XML report [#508](https://github.com/codeclimate/test-reporter/pull/508)

## v0.10.4 [(2022-8-9)](https://github.com/codeclimate/test-reporter/releases/tag/v0.10.4)

* [FIX] Updates Golang crypto dependency version [#497](https://github.com/codeclimate/test-reporter/pull/497)

## v0.10.3 [(2022-2-16)](https://github.com/codeclimate/test-reporter/releases/tag/v0.10.3)

* [FIX] Plasticize simplecov's formatter so it adapts to different reports structures  [#488](https://github.com/codeclimate/test-reporter/pull/488)

## v0.10.2 [(2021-12-7)](https://github.com/codeclimate/test-reporter/releases/tag/v0.10.2)

* [NEW] Add support for Github Actions [#481](https://github.com/codeclimate/test-reporter/pull/481)
* [FIX] Stop assigning coverage information to blank lines when using gocov [#480](https://github.com/codeclimate/test-reporter/pull/480)

## v0.10.1 [(2021-08-10)](https://github.com/codeclimate/test-reporter/releases/tag/v0.10.1)

* [NEW] Add support for Semaphore 2.0 [#473](https://github.com/codeclimate/test-reporter/pull/473)

## v0.10.0 [(2021-03-01)](https://github.com/codeclimate/test-reporter/releases/tag/v0.10.0)

* [NEW] Add support for Golang coverage generated using `-coverpkg` flag [#461](https://github.com/codeclimate/test-reporter/pull/461)

## v0.9.0 [(2020-12-15)](https://github.com/codeclimate/test-reporter/releases/tag/v0.9.0)

* [NEW] Add support for Simplecov `0.20.0` [#441](https://github.com/codeclimate/test-reporter/pull/441)
* [NEW] Add support for the llvm.coverage.json.export format. This is the format generated by the Swift Package Manager. [#439](https://github.com/codeclimate/test-reporter/pull/439)

## v0.8.0 [(2019-07-31)](https://github.com/codeclimate/test-reporter/releases/tag/v0.8.0)

* [NEW] Add `show-coverage` command [#424](https://github.com/codeclimate/test-reporter/pull/424)
* [NEW] Add Bitrise.io vars [#421](https://github.com/codeclimate/test-reporter/pull/421)
* [FIX] Docs [#405](https://github.com/codeclimate/test-reporter/pull/405)


## v0.7.0 [(2019-07-31)](https://github.com/codeclimate/test-reporter/releases/tag/v0.7.0)

* [NEW] Add support to xccov JSON report (Swift/Xcode 11) [#399](https://github.com/codeclimate/test-reporter/pull/399)


## v0.6.4 [(2019-01-22)](https://github.com/codeclimate/test-reporter/releases/tag/v0.6.4)

* [NEW] Updates release strategy to build a new Linux binary that uses
  `netcgo` [#382][]

[#382]: https://github.com/codeclimate/test-reporter/pull/382

## v0.6.3 [(2018-09-04)](https://github.com/codeclimate/test-reporter/releases/tag/v0.6.3)

* [NEW] Updates release strategy to build a new Linux binary that uses
  `netcgo` [#355][]

[#355]: https://github.com/codeclimate/test-reporter/pull/355

## v0.6.2 [(2018-07-31)](https://github.com/codeclimate/test-reporter/releases/tag/v0.6.2)

* [NEW] Add support for multiple JaCoCo source paths [#348][]

## v0.6.1 [(2018-07-30)](https://github.com/codeclimate/test-reporter/releases/tag/v0.6.1)

* [FIX] Support clover `path` attribute on file nodes in XML report [#349][]

[#349]: https://github.com/codeclimate/test-reporter/pull/349

## v0.6.0 [(2018-05-21)](https://github.com/codeclimate/test-reporter/releases/tag/v0.6.0)

* [FIX] Update `Gcov` formatter to report the correct source file paths [#338][]

[#338]: https://github.com/codeclimate/test-reporter/pull/338

## v0.5.2 [(2018-05-15)](https://github.com/codeclimate/test-reporter/releases/tag/v0.5.2)

* [FIX] Update `Cobertura` formatter to ignore invalid line numbers in a
  `cobertura.xml` file [#335][]

[#335]: https://github.com/codeclimate/test-reporter/pull/335

## v0.5.1 [(2018-03-19)](https://github.com/codeclimate/test-reporter/releases/tag/v0.5.1)

* [FIX] Fix bug with formatting JaCoCo test coverage [#318][]

[#318]: https://github.com/codeclimate/test-reporter/pull/318

## v0.5.0 [(2018-02-26)](https://github.com/codeclimate/test-reporter/releases/tag/v0.5.0)

* [NEW] Add flag to upload coverage insecurely [#310][]

[#310]: https://github.com/codeclimate/test-reporter/pull/310

## v0.4.5 [(2018-02-19)](https://github.com/codeclimate/test-reporter/releases/tag/v0.4.5)

* [FIX] Add partial automated support for Heroku CI builds [#305][]

[#305]: https://github.com/codeclimate/test-reporter/pull/305

## v0.4.4 [(2018-02-15)](https://github.com/codeclimate/test-reporter/releases/tag/v0.4.4)

* [FIX] Add support for Codeship CI environment variables [#300][]

## v0.4.3 [(2018-01-18)](https://github.com/codeclimate/test-reporter/releases/tag/v0.4.3)

[#300]: https://github.com/codeclimate/test-reporter/pull/300

* [FIX] Fix logging to not include extraneous newline, correct spelling [#288][]

[#288]: https://github.com/codeclimate/test-reporter/pull/288

## v0.4.2 [(2018-01-09)](https://github.com/codeclimate/test-reporter/releases/tag/v0.4.2)

* [FIX] Improved performance of the Cobertura, Gcov, and Jacoco formatters [#285][]

[#285]: https://github.com/codeclimate/test-reporter/pull/285

## v0.4.1 [(2018-01-08)](https://github.com/codeclimate/test-reporter/releases/tag/v0.4.1)

* [FIX] Improved performance of the LCOV formatter [#270][]

[#270]: https://github.com/codeclimate/test-reporter/pull/270

## v0.4.0 [(2017-12-19)](https://github.com/codeclimate/test-reporter/releases/tag/v0.4.0)

* [NEW] Add support for `excoveralls` json report (Elixir) [#278][]

[#278]: https://github.com/codeclimate/test-reporter/pull/278

## v0.3.4 [(2017-12-13)](https://github.com/codeclimate/test-reporter/releases/tag/v0.3.4)

* [FIX] `sum-coverage` command when merging source files with
        different coverage length but same blob ID [#272][]


upload/uploader.go
package upload

import (
	"bytes"
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/sirupsen/logrus"
	"github.com/codeclimate/test-reporter/formatters"
	"github.com/codeclimate/test-reporter/version"
	"github.com/pkg/errors"
)

type Uploader struct {
	ReporterID  string
	EndpointURL string
	BatchSize   int
	Input       io.Reader
	Insecure    bool
}

type ErrConflict struct {
	message string
}

func (e *ErrConflict) Error() string {
	return e.message
}

func (u Uploader) Upload() error {
	if u.ReporterID == "" {
		return errors.New("you must supply a CC_TEST_REPORTER_ID ENV variable or pass it via the -r flag")
	}

	rep := formatters.Report{
		SourceFiles: formatters.SourceFiles{},
	}

	err := json.NewDecoder(u.Input).Decode(&rep)
	if err != nil {
		return errors.WithStack(err)
	}

	testReport := NewTestReport(rep)

	pr, pw := io.Pipe()
	go func() {
		defer pw.Close()
		bb := &bytes.Buffer{}
		w := io.MultiWriter(pw, bb)
		err := json.NewEncoder(w).Encode(JSONWraper{Data: testReport})
		if err != nil {
			logrus.Error(err)
			return
		}
		logrus.Debug(bb.String())
	}()

	res, err := u.doRequest(pr, u.EndpointURL)
	if err != nil {
		switch err.(type) {
		case *ErrConflict:
			logrus.Warnf("%s, skipping upload", err.Error())
			return nil
		default:
			return errors.WithStack(err)
		}
	}

	batchLinks := struct {
		Links struct {
			PostBatch string `json:"post_batch"`
		} `json:"links"`
	}{}

	err = json.NewDecoder(res.Body).Decode(&batchLinks)
	if err != nil {
		return errors.WithStack(err)
	}

	postBatchURL, err := u.TransformPostBatchURL(batchLinks.Links.PostBatch)
	if err != nil {
		return errors.WithStack(err)
	}

	return u.SendBatches(testReport, postBatchURL)
}

func (u Uploader) TransformPostBatchURL(rawURL string) (string, error) {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return "", err
	}

	if u.Insecure {
		parsed.Scheme = "http"
	} else {
		parsed.Scheme = "https"
	}

	return parsed.String(), nil
}

func (u Uploader) SendBatches(rep *TestReport, url string) error {
	batch := [][]SourceFile{}

	pos := 0
	count := len(rep.SourceFiles) / u.BatchSize
	remainder := len(rep.SourceFiles) % u.BatchSize
	for i := 0; i < count; i++ {
		end := pos + u.BatchSize
		batch = append(batch, rep.SourceFiles[pos:end])
		pos = end
	}
	if remainder > 0 {
		batch = append(batch, rep.SourceFiles[pos:])
	}

	for i, b := range batch {
		pr, pw := io.Pipe()
		go func() {
			defer pw.Close()
			bb := &bytes.Buffer{}
			w := io.MultiWriter(pw, bb)
			err := json.NewEncoder(w).Encode(JSONWraper{
				Data: b,
				Meta: map[string]int{
					"current": i + 1,
					"total":   len(batch),
				},
			})
			if err != nil {
				logrus.Error(err)
				return
			}
			logrus.Debug(bb.String())
		}()
		_, err := u.doRequest(pr, url)
		if err != nil {
			return errors.WithStack(err)
		}
	}
	fmt.Println("Test report uploaded successfully to Code Climate")
	return nil
}

func (u Uploader) doRequest(in io.Reader, url string) (*http.Response, error) {
	c := http.Client{
		Transport: u.newTransport(),
		Timeout:   30 * time.Second,
	}

	req, err := u.newRequest(in, url)
	if err != nil {
		return nil, errors.WithStack(err)
	}

	logrus.Debugf("posting request to %s", url)
	res, err := c.Do(req)
	if err != nil {
		return res, errors.WithStack(err)
	}

	if res.StatusCode < 200 || res.StatusCode >= 300 {
		httpBody, err := ioutil.ReadAll(res.Body)
		if err != nil {
			return nil, errors.WithStack(err)
		}

		logrus.Debug(string(httpBody))
		errorMessage, err := getErrorMessage(httpBody)
		if err != nil {
			return nil, errors.WithStack(err)
		}

		if res.StatusCode == 409 {
			return nil, &ErrConflict{
				message: fmt.Sprintf("Conflict when uploading: %s", errorMessage),
			}
		}

		return res, fmt.Errorf("response from %s.\nHTTP %d: %s", url, res.StatusCode, errorMessage)
	}
	return res, nil
}

func (u Uploader) newTransport() (tr http.RoundTripper) {
	sslCertFile := os.Getenv("SSL_CERT_FILE")
	if sslCertFile == "" {
		return tr
	}

	caCert, err := ioutil.ReadFile(sslCertFile)
	if err != nil {
		log.Fatal(err)
	}
	caCertPool := x509.NewCertPool()
	caCertPool.AppendCertsFromPEM(caCert)
	return &http.Transport{
		TLSClientConfig: &tls.Config{
			RootCAs: caCertPool,
		},
	}
}

func (u Uploader) newRequest(in io.Reader, url string) (*http.Request, error) {
	req, err := http.NewRequest("POST", url, in)
	if err != nil {
		return nil, errors.WithStack(err)
	}
	req.Header.Set("User-Agent", fmt.Sprintf("TestReporter/%s (Code Climate, Inc.)", version.Version))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-CC-Test-Reporter-Id", u.ReporterID)
	req.Header.Set("Accept", "application/vnd.api+json")
	return req, err
}

type apiError struct {
	Detail string `json:"detail"`
}

type errorsResponse struct {
	Errors []apiError `json:"errors"`
}

func getErrorMessage(body []byte) (string, error) {
	var response = new(errorsResponse)
	err := json.Unmarshal(body, &response)
	if err != nil {
		return "", errors.WithStack(err)
	}

	var details []string

	for i := range response.Errors {
		details = append(details, response.Errors[i].Detail)
	}

	return strings.Join(details, ", "), err
}
